wsw/preprocessing.py

```diff
@@ -0,0 +1,258 @@
+
+    if save_to is not None:
+        if not os.path.exists(os.path.dirname(save_to)):
+            print(os.path.exists(os.path.dirname(save_to)))
+            print('Why are you here if the path exists???')
+            try:
+                os.makedirs(os.path.dirname(save_to))
+            except FileExistsError:
+                pass
+            with sf.SoundFile(save_to, 'w', sr, channels=1) as f:
+                f.write(audio)
+
+        if log is not None:
+            with open(log, 'a') as m:
+                m.write(os.path.basename(save_to) + '\n')
+    else:
+        return audio, sr
+
+
+def clip_all(fpath, save_to, length, sr=None, restart=False, log=None):
+    """
+    Clip or pad all files audio files found in the umbrella directory `fpath`
+    to a single desired length, then save to a new (or same) location. Files
+    will be saved in the `.wav` format.
+
+    :param fpath: The directory containing audio files in labelled folders. In
+    other words, the directory `fpath` should contain sub-directories where
+    each sub-directory is a unique label for the data within.
+    :param save_to: The location to save the clipped files too. Original
+        labelled file structure will be preserved.
+    :param length: The length to cut or pad variable length audio signal to.
+    :param sr: The sample rate of the audio being fingerprinted.
+    :param restart: If true then the `log` argument is mandatory. When true,
+        this method will look for a log file containing a list of already
+        fingerprinted files.
+    :param log: The path to the log which tracks the already completed files.
+        This argument is mandatory if `restart=True`.
+    """
+
+    if restart:
+        with open(log) as lg:
+            completed = set([line.strip() for line in lg])
+
+    folders = [d for d in os.scandir(fpath) if os.path.isdir(d.path) if not d.name.startswith('.')]
+    print(folders)
+
+    for i, folder in enumerate(folders):
+        dirname = folder.name
+        # print(f'Working on folder {dirname}. Overall progress: {int(100 * i / len(folders))}%   \r', end="")
+
+        if restart:
+            files = [f for f in os.scandir(folder) if (f.name not in completed) and os.path.isfile(f.path)]
+        else:
+            files = [f for f in os.scandir(folder) if os.path.isfile(f.path)]
+
+        fpaths = [f.path for f in files]
+        fnames = [f.name for f in files]
+
+        if files:
+            new_paths = [os.path.join(save_to, dirname, f) for f in fnames]
+
+            with Pool(os.cpu_count() - 1) as p:
+                n = len(files)
+                z = p.starmap(librosa.load, zip(fpaths, [sr] * n))
+                aud, srs = zip(*z)
+                p.starmap(clip_audio, zip(aud, [length] * n, srs, new_paths, [log] * n))
+
+    print('\nResizing of all audio files complete.')
+
+
+# def split_n_mix(audio, sr, thresh):
+#     return librosa.effects.remix(librosa.effects.split())
```

Comment:
Is this a helper function? If so make it protected. Also add in a description.

Reply:
I think I'm going to abandon that function for now. I was going to compress audio by removing silence/noise for the bird competition, then mixing the non-silent clips back in. I'll just delete it.